### PR TITLE
feat: ✨ upload data file page

### LINKED ISSUE #171: Develop the upload page for wireframes 2.2.1-2.2.3
Wireframes 2.2.1-2.2.3 are very similar and could probably be implemented at the same time.

### PR DESCRIPTION
**Pull request by lwjohnst86**
## Description

This PR finally adds the page for uploading the data files for #171.

- Please ignore the TODO items, since they will eventually need to be implemented but I can't do it all in this PR (plus depends on others work).
- Right now, the "feature" simply uploads the CSV file to the storage, which was already done by @philter87. The aim is that it joins existing data and updates the database with that data.
- The changes to the JavaScript is because of the formatter, not from me!!

## Related Issues

Closes #171 

## Testing

- [ ] Yes
- [ ] No, not needed (give a reason below)
- [x] No, I need help writing them

I haven't included any tests yet because right now it doesn't do anything that isn't already done. But once other functions have been implemented, those tests will need to be added.

## Reviewer Focus

This PR requires an in-depth review. 

## Checklist

For all PRs that are not general documentation

- [ ] Tests accompany or reflect changes to the code
- [ ] Tests ran and passed locally
- [x] Ran the linter and formatter
- [x] Build has passed locally
- [x] Relevant documentation has been updated

### CODE REVIEW

**Comment by lwjohnst86** [comment]
@seedcase-project/developers finished developing this! It's basic right now, but I have commented code in it that needs to eventually be implemented (but that I don't ccompletely know how to do right now, since it involves creating the database).

File: sprout/views/project_id_metadata_id_data_update.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        new_uploaded_file = get_uploaded_file(request)
+        file_metadata = FileMetadata.create_file_metadata(new_uploaded_file, table_id)
+        new_server_file = file_metadata.server_file_path
+        schema = get_schema(id=table_id)
+        data_update = read_csv_file(new_server_file, row_count=None)
+
+        # TODO: Implement these below later
+        # Inform if there are columns that exist in the uploaded data,
+        # that don't exist in the schema
+        # verify_headers(data, schema)
```
Comment (philter87) [inline sprout/views/project_id_metadata_id_data_update.py:37]: There is a function called validate_csv() in the file csv_validator, which returns a list of validation errors (column name mismatch or type mismatch)

This could be included in another PR if you like

File: sprout/views/project_id_metadata_id_data_update.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        #     # TODO: Not sure which engine to use. Alternative is "adbc"
+        #     # engine="sqlalchemy"
+        # )
+        # TODO: verify that database has been written to.
+
+        context = {
+            "table_name": table_metadata.name,
+            "upload_success": True,
+            "file_metadata": file_metadata,
+            "number_rows": count_rows(new_server_file),
```
Comment (philter87) [inline sprout/views/project_id_metadata_id_data_update.py:59]: write_database returns the row count, so this can eventually be used instead of reading the file again

File: sprout/templates/project-id-metadata-id-data-update.html [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+{% endblock %}
+
+{% block content %}
+  <div>
+    <h1 class="center-align">Upload File for {{ table_name }}</h1>
+
+    <!-- The file drop zone is included -->
+    {% include 'file-drop-zone.html' %}
+
+    {% if upload_success %}
```
Comment (philter87) [inline sprout/templates/project-id-metadata-id-data-update.html:16]: Nice with a snackbar :)

File: sprout/templates/project-id-metadata-id-data-update.html [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      <div class="snackbar active">
+        <div class="max">
+          <i>task_alt</i>
+          <span>Successfully uploaded <code class="file-ext">{{ file_metadata.original_file_name }}</code> and added {{ number_rows }} to {{ table_name }}</span>
+        </div>
+        <!-- TODO: Should this be here or as a button below the drop zone? -->
+        <a href="/metadata" class="inverse-link">Return to data view</a>
```
Comment (philter87) [inline sprout/templates/project-id-metadata-id-data-update.html:23]: Link is not working, but maybe that is okay if it is a TODO?

Comment (lwjohnst86) [reply sprout/templates/project-id-metadata-id-data-update.html:23]: It works now that @signekb's PR has been merged 😁

File: sprout/views/project_id_metadata_id_data_update.py (hunk 2 above)
Comment (lwjohnst86) [reply sprout/views/project_id_metadata_id_data_update.py:59]: Ooo, that's useful!

File: sprout/views/project_id_metadata_id_data_update.py (hunk 1 above)
Comment (lwjohnst86) [reply sprout/views/project_id_metadata_id_data_update.py:37]: Right! I will probably have to refactor it a bit to fit.

**Review by philter87 — DISMISSED** [review]

**Review by philter87 — DISMISSED** [review]

**Comment by lwjohnst86** [comment]
There's lots that still need to be done here, but I don't want this PR to get too big and there's other moving pieces that this code depends on.
